tests/test_cli.py

```diff
@@ -131,6 +131,114 @@ def test_cli_data_parsed_by_question_type(
+        ("float", "1.2", 1.2),
+        ("bool", "true", True),
+        ("bool", "false", False),
+        ("yaml", '"string"', "string"),
+        ("yaml", "1", 1),
+        ("yaml", "1.2", 1.2),
+        ("yaml", "true", True),
+        ("yaml", "null", None),
+        ("yaml", '["string", 1, 1.2, true, null]', ["string", 1, 1.2, True, None]),
+        ("json", '"string"', "string"),
+        ("json", "1", 1),
+        ("json", "1.2", 1.2),
+        ("json", "true", True),
+        ("json", "null", None),
+        ("json", '["string", 1, 1.2, true, null]', ["string", 1, 1.2, True, None]),
+    ],
+)
+def test_data_file_parsed_by_question_type(
+    tmp_path_factory: pytest.TempPathFactory, type_, answer, expected
+) -> None:
+    src, dst, local = map(tmp_path_factory.mktemp, ("src", "dst", "local"))
+    build_file_tree(
+        {
+            (src / "copier.yml"): (
+                f"""\
+                question:
+                    type: {type_}
+                """
+            ),
+            (src / "{{ _copier_conf.answers_file }}.jinja"): (
+                """\
+                # Changes here will be overwritten by Copier
+                {{ _copier_answers|to_nice_yaml }}
+                """
+            ),
+            (local / "data.yml"): yaml.dump({"question": answer}),
+        }
+    )
+
+    run_result = CopierApp.run(
+        [
+            "copier",
+            "copy",
+            f"--data-file={local / 'data.yml'}",
+            str(src),
+            str(dst),
+            "--quiet",
+        ],
+        exit=False,
+    )
+    assert run_result[1] == 0
+    answers = yaml.safe_load((dst / ".copier-answers.yml").read_text())
+    assert answers["_src_path"] == str(src)
+    assert answers["question"] == expected
+
+
+def test_data_and_data_file_mutually_exclusive(
+    tmp_path_factory: pytest.TempPathFactory,
+    capsys: pytest.CaptureFixture[str],
+):
+    src, dst, local = map(tmp_path_factory.mktemp, ("src", "dst", "local"))
+
+    build_file_tree(
+        {
+            (src / "copier.yml"): (
+                """\
+                question:
+                    type: str
+                """
+            ),
+            (src / "{{ _copier_conf.answers_file }}.jinja"): (
+                """\
+                # Changes here will be overwritten by Copier
+                {{ _copier_answers|to_nice_yaml }}
+                """
+            ),
+            (local / "data.yml"): yaml.dump({"question": "does not matter"}),
+        }
+    )
+
+    run_result = CopierApp.run(
+        [
+            "copier",
+            "copy",
+            f"--data-file={local / 'data.yml'}",
+            '--data=question="also does not matter"',
+            str(src),
+            str(dst),
+            "--quiet",
+        ],
+        exit=False,
+    )
+    assert run_result[1] == 2  # cli failure
+    out, _ = capsys.readouterr()
+    assert "Error: Given --data-file, the following are invalid ['--data']" in out
```

Comment:
It seems the order of the error for mutually exclusive flags isn't deterministic. Sometimes, this order of flags is printed, sometimes is swapped, i.e.:

```
assert "Error: Given --data, the following are invalid ['--data-file']" in out
```

Perhaps we can simply check for both cases:

```suggestion
    assert any(
        [
            "Error: Given --data-file, the following are invalid ['--data']" in out,
            "Error: Given --data, the following are invalid ['--data-file']" in out,
        ]
    )

```

Reply:
Happy to add this; it does seem like asserting an error condition of 2 and no data project created Shou be sufficient though?

Reply:
True, this seems to be a [`plumbum.cli.switches.SwitchCombinationError` which results in a return code 2](https://github.com/tomerfiliba/plumbum/blob/40fcd2708863d36fb570be4ca19c14d2457ab7b7/plumbum/cli/application.py#L626), but [Copier also uses return code 2 for `UnsafeTemplateError`](https://github.com/copier-org/copier/blob/ddafdee3ce162da7b889c44e11aad8b19c910136/copier/cli.py#L72-L74), so it's ambiguous. Resolving this return code collision is a [separate topic for discussion](https://github.com/copier-org/copier/issues/1328) by the way. So, I'd say it's safer to assert also the error message.